crates/datafusion_ext/src/stream.rs

```diff
@@ -0,0 +1,96 @@
+use std::any::Any;
+use std::fmt;
+use std::sync::{Arc, Mutex};
+
+use datafusion::arrow::datatypes::{Schema, SchemaRef};
+use datafusion::common::DataFusionError;
+use datafusion::execution::TaskContext;
+use datafusion::physical_plan::expressions::PhysicalSortExpr;
+use datafusion::physical_plan::metrics::{ExecutionPlanMetricsSet, MetricsSet};
+use datafusion::physical_plan::{
+    DisplayAs,
+    DisplayFormatType,
+    ExecutionPlan,
+    Partitioning,
+    SendableRecordBatchStream,
+    Statistics,
+};
+
+pub struct StreamExecPlan {
+    stream: Mutex<Option<SendableRecordBatchStream>>,
```

Comment:
Also, is this truly generic? Will other Executionplans use this? I guess i don't understand what this is for at all.

Reply:
to be honest the mutex/option is a bit wacky to me, but I got the pattern off of sean who did it [here](https://github.com/glaredb/glaredb/blob/main/crates/datasources/src/bson/stream.rs#L73-L93), because execution plans have to be send+sync, this is the thing that does it.

Reply:
Same comment as here. https://github.com/GlareDB/glaredb/pull/2613#discussion_r1483404663

I think we should have specialized `InsertExec`s for each datasource. This gives us the most flexibility later on, and avoids having to do this whole `Mutex<Option<>>` hack

Reply:
This strikes me as a premature optimization, and you'll still need to move the stream (or construct the stream in a kind of backwards way) and you'd still have to move the connection to MongoDB (in this case) which would probably be roughly the same kind of gross. 

In any case, unless there's a functional limitation, and given that the implementations can be relatively similar, I don't think this is worth the time.

Reply:
I wouldn't really consider it an "optimization" per-se. I think of it as more of a design decision. Custom execs provide more flexibility when we inevitably do need to optimize, or add other functionality that we don't have clear insight into right now. 

> you'd still have to move the connection to MongoDB (in this case) which would probably be roughly the same kind of gross.

`dyn ExecutionPlan` does not need to be send+sync like the record batch does. you can easily create a `MongoInsertExec` with just cloning the collection

```rust
#[derive(Debug)]
pub struct MongoInsertExec {
    input: Arc<dyn ExecutionPlan>,
    collection: Collection<RawDocumentBuf>,
}

impl MongoInsertExec {
    pub fn new(input: Arc<dyn ExecutionPlan>, collection: Collection<RawDocumentBuf>) -> Self {
        Self { input, collection }
    }
}


impl ExecutionPlan for MongoInsertExec {
// ...
}
// ...
```

```rust
    async fn insert_into(
        &self,
        state: &SessionState,
        input: Arc<dyn ExecutionPlan>,
        _overwrite: bool,
    ) -> DatafusionResult<Arc<dyn ExecutionPlan>> {
        Ok(Arc::new(MongoInsertExec::new(
            input,
            self.collection.clone(),
        )))
    }
```

Reply:
I guess another way of posing this is why do you think we should have a generic insert exec plan? to cut down on LOC? or are there other compelling reasons?

It seems that individual execs provide the most flexibility, makes the intent very clear, are pretty easy to implement, and dont require hacky workarounds like the `Mutex<Option<>>`

Reply:
I don't think the current design commits us to anything, and seems easy enough to refactor in the future when we better understand our needs. 

----

In your example, `MongoInsertExec` seems like it would need wrap something like the current execution plan, as the `inner` is also an execution plan, so an extra level of indirection seems easy enough to add in the future when we need to.